### ISSUE TITLE
Add a way to configure ODK for importing ontologies in their entirety (to support COCOONs)

### ISSUE THREAD

**Issue by cmungall** [INCATools/ontology-development-kit#411]
Often we want to make a mega-ontology that has zero axioms of our own, but instead merges a Coherent Collection of Orthogonal ONtologies (COCOON)

An example is the monarch ontology (config is in this repo)

here we don't want any terms.txt, we just want ALL of uberon, cl, etc

Maybe there is a way to do this I am missing - use filter option?

COCOON stack manifesto: https://docs.google.com/document/d/1Nr6mon6FEZAyvkAKC5OFmTLk5uANWzhoXUjxZVmOEJk/edit#

**Comment by matentzn** [comment]
This is already possible. There are two ways to do it:

1) If the import is considered a _proper part_ of the ontology, you can use components:

```
components:
  products:
    - filename: pato.owl
      source: http://purl.obolibrary.org/obo/pato/pato-base.owl
```

This means: CACOON has a component named pato.owl, and `source` is where to get it from. The advantage of using components is that the _signature of the components will be added to the ontology seed_.

2) If the import is considered an import, we recently introduced a new `module_type` called `mirror`. You can use it like this:

```
    - id: pno
      mirror_from: http://purl.obolibrary.org/obo/iao/pno.owl
      module_type: mirror
```

See here for a full example:
https://github.com/ufbmi/OMRSE/blob/ac8d48e19572f57436e2fb11e509b75244717764/src/ontology/omrse-odk.yaml#L43

**Comment by cmungall** [comment]
@wdduncan and I are trying this to make nmdco, we have specified two components, envo and po, as well as an ro import group

When we try this, 

 - nmdco-base.owl has envo and po merged in
 - nmdco.owl: only classes in RO

This is almost the opposite of what we expect. We want the main release file to have the components merged in. A base file doesn't make so much sense here as this is an "application ontology" at the terminal node in a dependency chain

**Comment by matentzn** [comment]
Why did the second, the `module_type: mirror` approach not work for you?

**Comment by cmungall** [comment]
Maybe we don't understand the terminology re component, proper-part

This is a COCOON use case

We want a simple merger or envo and po, so we can do queries on the two combined

is component not the right approach?

**Comment by matentzn** [comment]
Component means - proper part of ontology. In particular, the signature of the component is part of the signature of the base. "Module type mirror" means- the entire ontology is imported and treated as a module. In particular, the signature of the module is not part of the signature of the base. I think you want the latter!

**Comment by wdduncan** [comment]
sorry @matentzn I'm not understanding you :(  

I tried this:
```
components:
  products:
    - id: po
      filename: po.owl
      module_type: mirror
      source: http://purl.obolibrary.org/obo/po.owl
    - id: envo
      filename: envo.owl
      module_type: mirror
      source: http://purl.obolibrary.org/obo/envo.owl  
```
and go the same result; i.e., the nmdco-edit file was **not** a merge of the ontologies, and the `ro` file was given as an import.

Next I tried:
```
import_group:
  products:
    - id: ro
      module_type: slme
      slme_individuals: exclude
   - id: po
      filename: po.owl
      module_type: mirror
      source: http://purl.obolibrary.org/obo/po.owl
    - id: envo
      filename: envo.owl
      module_type: mirror
      source: http://purl.obolibrary.org/obo/envo.owl  
```
and received the error:
```
INVALID ELEMENT ERROR "http://www.goosocean.org/components/com_oe/oe.php?task=download&id=34507&version=1.0&lang=1&format=1" contains invalid characters
For details see: http://robot.obolibrary.org/errors#invalid-element-error
Use the -vvv option to show the stack trace.
Use the --help option to see usage information.
Makefile:239: recipe for target 'imports/envo_import.owl' failed
```

**Comment by matentzn** [comment]
If I understand your use case correctly, that is what you want:

```
import_group:
  products:
    - id: ro
      module_type: slme
      slme_individuals: exclude
   - id: po
      module_type: mirror
   - id: envo
      module_type: mirror
```

**Comment by matentzn** [comment]
Ignore this, this is just for your information:

```
- id: envo
     filename: envo.owl
     module_type: mirror
     source: http://purl.obolibrary.org/obo/envo.owl  
```

There is no `filename` element, and the `source` element is called `mirror_from`.

**Comment by wdduncan** [comment]
@matentzn This is my `import_group`:
```
import_group:
  products:
    - id: ro
      module_type: slme
      slme_individuals: exclude
    - id: po
      module_type: mirror
    - id: envo
      module_type: mirror
```
I received the error:
```python
2021-05-05 13:19:27,762 ERROR org.obolibrary.robot.IOHelper - Input ontology contains 1 triple(s) that could not be parsed:
 - <https://www.wikidata.org/wiki/Q2306597> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> _:genid-nodeid-node1f4uatpnhx4483.

INVALID ELEMENT ERROR "http://www.goosocean.org/components/com_oe/oe.php?task=download&id=34507&version=1.0&lang=1&format=1" contains invalid characters
For details see: http://robot.obolibrary.org/errors#invalid-element-error
Use the -vvv option to show the stack trace.
Use the --help option to see usage information.
Makefile:239: recipe for target 'imports/envo_import.owl' failed
```
I did not have a `components` group. Was there supposed to one?

**Comment by matentzn** [comment]
This seems unrelated to ODK - something about envo appears broken?

can you try to run just:

```
cd target/youront/src/ontology
sh run.sh make refresh-envo ROBOT= robot -vvv
```

**Comment by matentzn** [comment]
Looks good your file!

**Comment by wdduncan** [comment]
Sorry ... I'm out of my depth here.

I ran `sh run.sh make refresh-envo ROBOT="robot -vvv"`

It produced the error:
```python
INVALID ELEMENT ERROR "http://www.goosocean.org/components/com_oe/oe.php?task=download&id=34507&version=1.0&lang=1&format=1" contains invalid characters
For details see: http://robot.obolibrary.org/errors#invalid-element-error
java.io.IOException: errors#INVALID ELEMENT ERROR "http://www.goosocean.org/components/com_oe/oe.php?task=download&id=34507&version=1.0&lang=1&format=1" contains invalid characters
	at org.obolibrary.robot.IOHelper.saveOntologyFile(IOHelper.java:1724)
	at org.obolibrary.robot.IOHelper.saveOntology(IOHelper.java:838)
	at org.obolibrary.robot.CommandLineHelper.maybeSaveOutput(CommandLineHelper.java:671)
	at org.obolibrary.robot.AnnotateCommand.execute(AnnotateCommand.java:281)
	at org.obolibrary.robot.CommandManager.executeCommand(CommandManager.java:248)
	at org.obolibrary.robot.CommandManager.execute(CommandManager.java:192)
	at org.obolibrary.robot.CommandManager.main(CommandManager.java:139)
	at org.obolibrary.robot.CommandLineInterface.main(CommandLineInterface.java:60)
Makefile:239: recipe for target 'imports/envo_import.owl' failed
make[1]: *** [imports/envo_import.owl] Error 1
make[1]: Leaving directory '/work/src/ontology'
Makefile:253: recipe for target 'refresh-envo' failed
make: *** [refresh-envo] Error 2
```

I don't know what `"http://www.goosocean.org/components/com_oe/oe.php?task=download&id=34507&version=1.0&lang=1&format=1" ` is.

**Comment by matentzn** [comment]
It's a problem with ENVO that coincides with a feature of ODK. I will try to find a workaround for you.

**Comment by cmungall** [comment]
Wait shouldn't we fix ENVO?

```owl
$ curl -L -s $OBO/envo.owl | grep goosocean
        <oboInOwl:hasDbXref>http://www.goosocean.org/components/com_oe/oe.php?task=download&amp;id=34494&amp;version=1.0&amp;lang=1&amp;format=1</oboInOwl:hasDbXref>
        <oboInOwl:hasDbXref>http://www.goosocean.org/index.php?option=com_oe&amp;task=viewDocumentRecord&amp;docID=17464</oboInOwl:hasDbXref>
        <oboInOwl:hasDbXref>http://www.goosocean.org/index.php?option=com_oe&amp;task=viewDocumentRecord&amp;docID=17465</oboInOwl:hasDbXref>
        <oboInOwl:hasDbXref rdf:resource="http://www.goosocean.org/components/com_oe/oe.php?task=download&amp;id=34507&amp;version=1.0&amp;lang=1&amp;format=1"/>
        <oboInOwl:inSubset rdf:resource="http://www.goosocean.org/components/com_oe/oe.php?task=download&amp;id=34507&amp;version=1.0&amp;lang=1&amp;format=1"/>
        <oboInOwl:hasDbXref rdf:resource="http://www.goosocean.org/components/com_oe/oe.php?task=download&amp;id=34498&amp;version=1.0&amp;lang=1&amp;format=1"/>
```

is this really intentional?

**Comment by wdduncan** [comment]
Filed [ticket](https://github.com/EnvironmentOntology/envo/issues/1123) on ENVO.

**Comment by matentzn** [comment]
Thanks @wdduncan I proposed a fix here:
https://github.com/EnvironmentOntology/envo/pull/1124

**Comment by wdduncan** [comment]
Thanks @matentzn ! I was not aware that should have been an `xref`.

However AFIK, a new release of ENVO need to be produced before we can test. Right? Or can I use the `envo-edit` file from the ENVO repo?

Maybe @cmungall or @kaiiam can help get a new release pushed out?

**Comment by matentzn** [comment]
I made the ODK robust against this bug now, so if you `docker pull obolibrary/odkfull` you should be able to run everything.

**Comment by kaiiam** [comment]
@matentzn to the rescue as usual 🥇

**Comment by kaiiam** [comment]
> I was not aware that should have been an xref.

@wdduncan please consult our https://github.com/EnvironmentOntology/envo/wiki/ENVO-Robot-template-and-merge-workflow, specifically the Cross references section. I've tried to explain there the standards APs to use.

**Comment by wdduncan** [comment]
Thanks @kaiiam 

The issue had to with the **value** of the `in_subset` annotation, not the annotation itself :)

That is why I filed the ticket on the envo tracker :)

**Comment by kaiiam** [comment]
> The issue had to with the value of the in_subset annotation, not the annotation itself :)

Yes fair enough, we just don't put subset tags on definitions in ENVO like we do with dbxrefs. Subsets are a controlled set of APs.

**Comment by wdduncan** [comment]
@matentzn I verified that this works:
```
import_group:
  products:
    - id: ro
      module_type: slme
      slme_individuals: exclude
    - id: po
      module_type: mirror
    - id: envo
      module_type: mirror
```
It produces an `nmdco-edit.owl` file with the specified imports.

@cmungall did you want the output to be an `nmdco-edit.owl` that was a **merge** of `envo` and `po`?

**Comment by matentzn** [comment]
Awesome! Happy. The ODK will create a merge in any case when you run a release :)

**Comment by wdduncan** [comment]
Thanks for you help @matentzn !
I just wasn't sure if that is what @cmungall wanted ...

I am, however, confused about the intention of the
```
components:
  products:
    ....
```
specification. That is, how it differs from imports.

**Comment by matentzn** [comment]
Just think of it like this:

- imports are subsets of external ontologies, similar to dependencies.
- components are proper parts of the actual ontology, basically like an extension of the edit file. In essence, anything in a component could also be in the edit file

This has consequences for downstream tasks. For example: 
1. components are part of the ontology base file. imports are not
2. components signature will be used to extract imports - imports signatures do not affect (other) imports.

**Comment by wdduncan** [comment]
Thanks @matentzn !
Do you want to close this ticket?
